witness/internal/witness/witness.go

```diff
@@ -0,0 +1,151 @@
+package witness
+
+import (
+	"fmt"
+
+	"github.com/google/trillian-examples/formats/log"
+	"github.com/google/trillian/merkle/hashers"
+	"github.com/google/trillian/merkle/logverifier"
+	"golang.org/x/mod/sumdb/note"
+)
+
+type Chkpt struct {
+	Parsed	*log.Checkpoint
+	Raw	[]byte
+}
+
+type ChkptStorage interface {
+	// GetLatest returns the latest checkpoint for a given log.
+	GetLatest(logPK string) (*Chkpt, error)
+
+	// SetCheckpoint adds a checkpoint to the storage for a given log.
+	SetCheckpoint(logPK string, c Chkpt) error
+}
+
+type Witness struct {
+	db       ChkptStorage
+	Signer	 note.Signer
+	Hasher   hashers.LogHasher
+	LogV	 logverifier.LogVerifier
+	SigVs	 []note.Verifier
+}
+
+// NewWitness creates a new witness, which initially has no logs to follow.
+func NewWitness(h hashers.LogHasher, db ChkptStorage, sk string) (*Witness, error) {
+	ns, err := note.NewSigner(sk)
+	if err != nil {
+		return nil, err
+	}
+	return &Witness{
+		db:       db,
+		Signer:	  ns,
+		Hasher:   h,
+		LogV:	  logverifier.New(h),
+		SigVs:    []note.Verifier{},
+	}, nil
+}
+
+// parse verifies the checkpoint under the given list of possible public keys.  
+// It returns the parsed checkpoint and the list of public keys under which the 
+// signature(s) verified.
+func (w *Witness) parse(chkptRaw []byte, verifiers []note.Verifier) (*log.Checkpoint, []string, error) {
+	n, err := note.Open(chkptRaw, note.VerifierList(verifiers...))
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to verify checkpoint: %w", err)
+	}
+	chkpt := &log.Checkpoint{}
+	_, err = chkpt.Unmarshal([]byte(n.Text))
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to unmarshal new checkpoint: %w", err)
+	}
+	var keys []string
+	for _, sig := range n.Sigs {
+	    keys = append(keys, sig.Name)
+	}
+	return chkpt, keys, nil
+}
+
+// GetLatest gets the latest checkpoint for a given log, which should be
+// consistent with all other checkpoints for the same log.  It also signs it
+// under the witness' key.
+func (w *Witness) GetLatest(logPK string) ([]byte, error) {
+	chkpt, err := w.db.GetLatest(logPK)
+	if err != nil {
+		return nil, err
+	}
+	// Add the witness' signature to the checkpoint.
+	// TODO find a way so we don't have to verify the signatures again.
+	n, err := note.Open(chkpt.Raw, note.VerifierList(w.SigVs...))
+	if err != nil {
+		return nil, fmt.Errorf("failed to verify checkpoint: %w", err)
+	}
+	cosigned, err := note.Sign(n, w.Signer)
+	if err != nil {
+		return nil, err
+	}
+	return cosigned, nil
+}
+
+// Update updates latest checkpoint if `to` matches the current version (`from`).
+func (w *Witness) Update(fromRaw, toRaw []byte, proof log.Proof) error {
```

Comment:
Do we need `fromRaw` - it's implicitly going to have to be equivalent to the latest checkpoint we've seen from the log or we'll return an `ErrStale`?

Edit: below doesn't do what I thought it would - added a comment there too :)

Reply:
I agree the way it is now makes giving this as input a bit pointless, but I was worried about removing it because it's not clear what would happen in a real scenario.  I guess the "follower"/"feeder" would call `GetLatest`, query the log for a consistency proof between the returned checkpoint and the log's latest one, and then give just the latest checkpoint and proof as input to `Update`?

Reply:
Okay, fromRaw is gone now.